# Eat Safe, Love

## Part 1: Database and Jupyter Notebook Set Up

##This step is established----------------

Import the data provided in the `establishments.json` file from your Terminal. Name the database `uk_food` and the collection `establishments`.

Within this markdown cell, copy the line of text you used to import the data from your Terminal. This way, future analysts will be able to repeat your process.

e.g.: Import the dataset with `mongoimport --type json -d uk_food -c establishments --drop --jsonArray establishments.json`

##attached screenshot docs - <<Module 12 nosql-challenge - screen shots v1>>

In [47]:
# Import dependencies
from pymongo import MongoClient
from pprint import pprint
import pandas as pd

In [48]:
# Create an instance of MongoClient
mongo = MongoClient(port=27017)

In [49]:
# confirm that our new database was created
print(mongo.list_database_names())

['admin', 'config', 'local', 'met', 'uk_food']


In [50]:
# assign the uk_food database to a variable name
db = mongo['uk_food']

In [51]:
# review the collections in our new database
print(db.list_collection_names())

['establishments']


In [52]:
# review a document in the establishments collection
pprint(db.establishments.find_one())

{'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'Distance': 4623.972328074718,
 'LocalAuthorityCode': '511',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'NewRatingPending': True,
 'Phone': '',
 'PostCode': 'SE18 7DY',
 'RatingValue': None,
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('64b4631e749677635e6ce01e'),
 'geocode': {'latitude': 51.490142, 'longitude': 0.08384},
 'scores': {'ConfidenceInManagement': '', 'Hygiene': '', 'Structural': ''}}


In [53]:
# assign the collection to a variable
establishments = db['establishments']

## Part 2: Update the Database

1. An exciting new halal restaurant just opened in Greenwich, but hasn't been rated yet. The magazine has asked you to include it in your analysis. Add the following restaurant "Penang Flavours" to the database.

In [54]:
# Create a dictionary for the new restaurant data
new_restaurant = {
    "BusinessName":"Penang Flavours",
    "BusinessType":"Restaurant/Cafe/Canteen",
    "BusinessTypeID":"",
    "AddressLine1":"Penang Flavours",
    "AddressLine2":"146A Plumstead Rd",
    "AddressLine3":"London",
    "AddressLine4":"",
    "PostCode":"SE18 7DY",
    "Phone":"",
    "LocalAuthorityCode":"511",
    "LocalAuthorityName":"Greenwich",
    "LocalAuthorityWebSite":"http://www.royalgreenwich.gov.uk",
    "LocalAuthorityEmailAddress":"health@royalgreenwich.gov.uk",
    "scores":{
        "Hygiene":"",
        "Structural":"",
        "ConfidenceInManagement":""
    },
    "SchemeType":"FHRS",
    "geocode":{
        "longitude":"0.08384000",
        "latitude":"51.49014200"
    },
    "RightToReply":"",
    "Distance":4623.9723280747176,
    "NewRatingPending":True
}

In [55]:
# Insert the new restaurant into the collection
db.establishments.insert_one(new_restaurant)

In [56]:
# Check that the new restaurant was inserted
query = {'BusinessName': 'Penang Flavours'}
results = db.establishments.find(query)
for result in results:
    pprint(result)

{'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'Distance': 4623.972328074718,
 'LocalAuthorityCode': '511',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'NewRatingPending': True,
 'Phone': '',
 'PostCode': 'SE18 7DY',
 'RatingValue': None,
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('64b4631e749677635e6ce01e'),
 'geocode': {'latitude': 51.490142, 'longitude': 0.08384},
 'scores': {'ConfidenceInManagement': '', 'Hygiene': '', 'Structural': ''}}
{'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': '',
 'Distance': 46

2. Find the BusinessTypeID for "Restaurant/Cafe/Canteen" and return only the `BusinessTypeID` and `BusinessType` fields.

In [57]:
# Find the BusinessTypeID for "Restaurant/Cafe/Canteen" and return only the BusinessTypeID and BusinessType fields
query = {"BusinessType": "Restaurant/Cafe/Canteen"}
fields = {"BusinessTypeID":1, "BusinessType":1}
limit = 1

# Cast the results as a list and save the results to a variable
data = establishments.find(query, fields).limit(limit)

# Pretty print the results
df = pd.DataFrame(data)
df.head()

_id             BusinessType  BusinessTypeID
0  64b4631e749677635e6ce01e  Restaurant/Cafe/Canteen               1

3. Update the new restaurant with the `BusinessTypeID` you found.

In [58]:
# Update the new restaurant with the correct BusinessTypeID
establishments.update_one({'BusinessName': 'Penang Flavours'}, {'$set': {'BusinessTypeID': 1}})

In [59]:
# Confirm that the new restaurant was updated
query = {'BusinessName': 'Penang Flavours'}
results = db.establishments.find(query)
for result in results:
    pprint(result)

{'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'Distance': 4623.972328074718,
 'LocalAuthorityCode': '511',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'NewRatingPending': True,
 'Phone': '',
 'PostCode': 'SE18 7DY',
 'RatingValue': None,
 'RightToReply': '',
 'SchemeType': 'FHRS',
 '_id': ObjectId('64b4631e749677635e6ce01e'),
 'geocode': {'latitude': 51.490142, 'longitude': 0.08384},
 'scores': {'ConfidenceInManagement': '', 'Hygiene': '', 'Structural': ''}}
{'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': '',
 'Distance': 46

4. The magazine is not interested in any establishments in Dover, so check how many documents contain the Dover Local Authority. Then, remove any establishments within the Dover Local Authority from the database, and check the number of documents to ensure they were deleted.

In [60]:
# Find how many documents have LocalAuthorityName as "Dover"
query = {"LocalAuthorityName": "Dover"}

# Cast the results as a list and save the results to a variable
data = establishments.find(query)

# Pretty print the results
df = pd.DataFrame(data)
print(df.shape)
df.head()

(994, 28)


_id   FHRSID  ChangesByServerID  \
0  64b469d3fb4578cd17f8c216   254719                  0   
1  64b469d3fb4578cd17f8c217  1034540                  0   
2  64b469d3fb4578cd17f8c218   254250                  0   
3  64b469d3fb4578cd17f8c219   551803                  0   
4  64b469d3fb4578cd17f8c21a   632212                  0   

  LocalAuthorityBusinessID        BusinessName  \
0             PI/000069980   Refreshment Kiosk   
1             PI/000078691  The Coastguard Inn   
2             PI/000066174     The Pines Calyx   
3             PI/000070948        The Tea Room   
4             PI/000043474         Lenox House   

                        BusinessType  BusinessTypeID       AddressLine1  \
0            Restaurant/Cafe/Canteen               1            The Bay   
1                  Pub/bar/nightclub            7843            The Bay   
2            Other catering premises            7841   The Pines Garden   
3            Restaurant/Cafe/Canteen               1   The Pines Garden   
4  Hotel/bed & breakfast/guest house            7842  27 Granville Road   

       AddressLine2      AddressLine3  ...     LocalAuthorityWebSite  \
0  St Margarets Bay              Kent  ...  http://www.dover.gov.uk/   
1  St Margarets Bay              Kent  ...  http://www.dover.gov.uk/   
2        Beach Road  St Margarets Bay  ...  http://www.dover.gov.uk/   
3        Beach Road  St Margarets Bay  ...  http://www.dover.gov.uk/   
4  St Margarets Bay              Kent  ...  http://www.dover.gov.uk/   

      LocalAuthorityEmailAddress  \
0  publicprotection@dover.gov.uk   
1  publicprotection@dover.gov.uk   
2  publicprotection@dover.gov.uk   
3  publicprotection@dover.gov.uk   
4  publicprotection@dover.gov.uk   

                                              scores SchemeType  \
0  {'Hygiene': 0, 'Structural': 5, 'ConfidenceInM...       FHRS   
1  {'Hygiene': 0, 'Structural': 5, 'ConfidenceInM...       FHRS   
2  {'Hygiene': 0, 'Structural': 0, 'ConfidenceInM...       FHRS   
3  {'Hygiene': 0, 'Structural': 0, 'ConfidenceInM...       FHRS   
4  {'Hygiene': 0, 'Structural': 0, 'ConfidenceInM...       FHRS   

                                             geocode RightToReply  \
0  {'longitude': '1.387974', 'latitude': '51.1522...                
1  {'longitude': '1.387974', 'latitude': '51.1522...                
2  {'longitude': '1.383298', 'latitude': '51.1481...                
3  {'longitude': '1.383298', 'latitude': '51.1481...                
4  {'longitude': '1.385585', 'latitude': '51.1524...                

      Distance NewRatingPending  \
0  4587.347175            False   
1  4587.347175            False   
2  4587.362403            False   
3  4587.362403            False   
4  4587.440907            False   

                                                meta  \
0  {'dataSource': None, 'extractDate': '0001-01-0...   
1  {'dataSource': None, 'extractDate': '0001-01-0...   
2  {'dataSource': None, 'extractDate': '0001-01-0...   
3  {'dataSource': None, 'extractDate': '0001-01-0...   
4  {'dataSource': None, 'extractDate': '0001-01-0...   

                                               links  
0  [{'rel': 'self', 'href': 'https://api.ratings....  
1  [{'rel': 'self', 'href': 'https://api.ratings....  
2  [{'rel': 'self', 'href': 'https://api.ratings....  
3  [{'rel': 'self', 'href': 'https://api.ratings....  
4  [{'rel': 'self', 'href': 'https://api.ratings....  

[5 rows x 28 columns]

## Result results of "LocalAuthorityName": "Dover"}
# (994, 28)
# is saved as a screen shot doc-----------<<Module 12 nosql-challenge - screen shots v1>>
# Before running next step command for -----Deleteting 

In [68]:
# Delete all documents where LocalAuthorityName is "Dover"
query = {"LocalAuthorityName": "Dover"}
establishments.delete_many(query)

In [69]:
# Check if any remaining documents include Dover
results = establishments.find(query)

for result in results:
    print(result)

In [62]:
# Check that other documents remain with 'find_one'
establishments.find_one()

{'_id': ObjectId('64b4631e749677635e6ce01e'),
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'PostCode': 'SE18 7DY',
 'Phone': '',
 'LocalAuthorityCode': '511',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'scores': {'Hygiene': '', 'Structural': '', 'ConfidenceInManagement': ''},
 'SchemeType': 'FHRS',
 'geocode': {'longitude': 0.08384, 'latitude': 51.490142},
 'RightToReply': '',
 'Distance': 4623.972328074718,
 'NewRatingPending': True,
 'RatingValue': None}

5. Some of the number values are stored as strings, when they should be stored as numbers.

Use `update_many` to convert `latitude` and `longitude` to decimal numbers.

In [70]:
# Change the data type from String to Decimal for longitude and latitude
establishments.update_many({}, [ {'$set': { "geocode.longitude" : {'$toDouble': "$geocode.longitude"},
                                               "geocode.latitude" : {'$toDouble': "$geocode.latitude"}}}]) 

establishments.find_one()

{'_id': ObjectId('64b4631e749677635e6ce01e'),
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'PostCode': 'SE18 7DY',
 'Phone': '',
 'LocalAuthorityCode': '511',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'scores': {'Hygiene': '', 'Structural': '', 'ConfidenceInManagement': ''},
 'SchemeType': 'FHRS',
 'geocode': {'longitude': 0.08384, 'latitude': 51.490142},
 'RightToReply': '',
 'Distance': 4623.972328074718,
 'NewRatingPending': True,
 'RatingValue': None}

Use `update_many` to convert `RatingValue` to integer numbers.

In [71]:
# Set non 1-5 Rating Values to Null
non_ratings = ["AwaitingInspection", "Awaiting Inspection", "AwaitingPublication", "Pass", "Exempt"]
establishments.update_many({"RatingValue": {"$in": non_ratings}}, [ {'$set':{ "RatingValue" : None}} ])

In [72]:
# Change the data type from String to Integer for RatingValue
establishments.update_many({}, [ {'$set': { "RatingValue" : {'$toInt': "$RatingValue"}}}]) 

In [73]:
# Check that the coordinates and rating value are now numbers
establishments.find_one()

{'_id': ObjectId('64b4631e749677635e6ce01e'),
 'BusinessName': 'Penang Flavours',
 'BusinessType': 'Restaurant/Cafe/Canteen',
 'BusinessTypeID': 1,
 'AddressLine1': 'Penang Flavours',
 'AddressLine2': '146A Plumstead Rd',
 'AddressLine3': 'London',
 'AddressLine4': '',
 'PostCode': 'SE18 7DY',
 'Phone': '',
 'LocalAuthorityCode': '511',
 'LocalAuthorityName': 'Greenwich',
 'LocalAuthorityWebSite': 'http://www.royalgreenwich.gov.uk',
 'LocalAuthorityEmailAddress': 'health@royalgreenwich.gov.uk',
 'scores': {'Hygiene': '', 'Structural': '', 'ConfidenceInManagement': ''},
 'SchemeType': 'FHRS',
 'geocode': {'longitude': 0.08384, 'latitude': 51.490142},
 'RightToReply': '',
 'Distance': 4623.972328074718,
 'NewRatingPending': True,
 'RatingValue': None}

In [74]:
##End